### PR TITLE
api: direct response

### PR DESCRIPTION
**Pull request by arkodg**
Relates to https://github.com/envoyproxy/gateway/issues/2714

### CODE REVIEW

File: api/v1alpha1/shared_types.go [hunk 1]
```diff
@@ -663,7 +663,14 @@ type CustomResponse struct {
...
 	// Body of the Custom Response
-	Body CustomResponseBody `json:"body"`
+	//
+	// +optional
+	Body *CustomResponseBody `json:"body,omitempty"`
+
+	// Status Code of the HTTP response
+	//
+	// +optional
+	StatusCode *string `json:"statusCode,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/shared_types.go:673]: this is optional? what will happen if both `body` and `statusCode` is nil?

Comment (arkodg) [reply api/v1alpha1/shared_types.go:673]: Empty status code in this context will be a 200

Can add a cel validation to make sure at least one field must be set

Comment (zirain) [reply api/v1alpha1/shared_types.go:673]: > Empty status code for direct response will be a 200.
For response override it is the existing status code

Should address this? maybe not share the struct for better doc(I'm not sure about this)?

Comment (arkodg) [reply api/v1alpha1/shared_types.go:673]: Yup battling that same thought rn :)

**Review by Xunzhuo — DISMISSED** [review]
let us get this in

File: api/v1alpha1/shared_types.go (hunk 1 above)
Comment (zhaohuabing) [reply api/v1alpha1/shared_types.go:673]: Prefer two different types for direct response and override response for below reasons:
* For better doc as @zirain pointed out
* They may have different fields, for example, override response could add/remove headers

**Review by zirain — CHANGES_REQUESTED** [review]
wait https://github.com/envoyproxy/gateway/pull/4334#discussion_r1779996658

**Review by zirain — APPROVED** [review]

File: api/v1alpha1/shared_types.go [hunk 2]
```diff
@@ -669,7 +669,14 @@ type CustomResponse struct {
...
```
Comment (zhaohuabing) [inline api/v1alpha1/shared_types.go:679]: Nit: The `StatusCodeMatch` uses int for StatusCode, should we also use int here?

**Review by zhaohuabing — APPROVED** [review]

File: api/v1alpha1/shared_types.go (hunk 2 above)
Comment (arkodg) [reply api/v1alpha1/shared_types.go:679]: oops, I'll just delete this field for now

**Review by zirain — APPROVED** [review]

**Review by zhaohuabing — APPROVED** [review]

**Pull request merged**